### ISSUE TITLE
[HOLD Payment card / Subscription project] Add SCA support in Transfer Owner flow

### ISSUE THREAD

**Issue by amyevans** [Expensify/App#42490]
Per [this discussion](https://docs.google.com/document/d/1jXV_vIsOGZuL0dMiuXDWRETAFRX64ODVAhWPtiuFXT8/edit?pli=1&disco=AAABNev_yVY) in the [Payment card and subscription management in NewDot design doc](https://docs.google.com/document/d/1jXV_vIsOGZuL0dMiuXDWRETAFRX64ODVAhWPtiuFXT8/), we need to support adding a GBP payment card in the "Transfer Owner" flow.

### To test
1. As the owner of a Collect policy, add a new admin
2. Log into NewDot as that admin
3. Navigate to Account Settings > Workspaces > [workspace] > Members > click on the row of the _current_ owner > Transfer Owner

Currently, when transferring ownership of a policy in NewDot, the `AddBillingCardAndRequestPolicyOwnerChange` API command is called. That command is calling `UserAPI::addPaymentCard`, which doesn't account for “Strong Customer Authentication” aka SCA requirements (which are relevant for GBP-currency payment cards).

Once SCA support is built out for NewDot in the Payment card project, we should use it as a base to extend to the Transfer Owner flow.

**Comment by blimpich** [comment]
Still waiting for SCA to work in the subscription page, making progress.

**Comment by blimpich** [comment]
Picking this up.

**Comment by blimpich** [comment]
Hmm okay, looked at this this afternoon. Notes for myself:
- [this](https://github.com/Expensify/App/blob/55a29dd95d9753be05ddab40c79b79186ce05453/src/pages/workspace/members/WorkspaceOwnerChangeWrapperPage.tsx#L75) is where we decide whether to go through with the transfer or to ask for the user to add a card in App. Depends on whether `RequestWorkspaceOwnerChange` returns an error or not.
- It seems like we have a couple options here, but I think the best is gonna be a purely frontend change. I think we can get rid of `AddBillingCardAndRequestPolicyOwnerChange` entirely, which is nice since that command is not 1:1:1

**Current thinking:** lets just do the same thing we did with the add payment card form here. We can use [addSubscriptionPaymentCard()](https://github.com/Expensify/App/blob/55a29dd95d9753be05ddab40c79b79186ce05453/src/libs/actions/PaymentMethods.ts#L210) in place of addBillingCardAndRequestPolicyOwnerChange() right [here](https://github.com/Expensify/App/blob/55a29dd95d9753be05ddab40c79b79186ce05453/src/pages/workspace/members/WorkspaceOwnerPaymentCardForm.tsx#L73). If its a non-GBP currency it'll go through as normal, and if it is GBP it'll start the setupIntent flow and will give us an `authenticationLink` that we can use to render a [CardAuthenticationModal component](https://github.com/Expensify/App/blob/55a29dd95d9753be05ddab40c79b79186ce05453/src/pages/settings/Subscription/CardAuthenticationModal/index.tsx#L19). Then, in `WorkspaceOwnerPaymentCardForm.tsx` we can add two `useEffect` hooks, one to wait for the onyx key `fundList` to update, and one to waits for a `3DS-authentication-complete` postMessage. The hook that is waiting on the 3DS post message will do basically the exact same thing that we do in add Payment card (see [here](https://github.com/Expensify/App/blob/55a29dd95d9753be05ddab40c79b79186ce05453/src/pages/settings/Subscription/CardAuthenticationModal/index.tsx#L38)), and the hook that waits for the fundList to update will check if the fundList has updated, if it has it'll check to make sure the billing card doesn't have any errors associated with it, and if it doesn't then it'll call the existing `RequestWorkspaceOwnerChange` command to change ownership.

I'm not actually even sure that we need to do that last part where we call `RequestWorkspaceOwnerChange` since it looks like we might currently be making duplicate calls. Before assigning this out to anyone I'm going to look into it more. But yeah, on initial inspection I think this can be all frontend and we can maybe even get rid of a whole command on the backend.

Follow up items for @blimpich to look into:
- [ ] does the transfer ownership flow break if we delete [this line](https://github.com/Expensify/Web-Expensify/blob/main/lib/PolicyAPI.php#L5984) in web? Seems like it shouldn't, seems like we might be calling the web api `PolicyAPI::transferOwnership` twice right now in this flow, which would make it redundant

**Comment by blimpich** [comment]
Yeah that line seems to be redundant, so I don't think we need `AddBillingCardAndRequestPolicyOwnerChange` at all.

**Comment by blimpich** [comment]
Okay I made a proof of concept branch just to see if this idea would even work, and it does! I can transfer the owner by just waiting to see if the fund list updates and then triggering a RequestOwner. However after thinking about this more I think this might be violating our 1:1:1 philosophy.

**Comment by blimpich** [comment]
Ran out of time but I will ask in engineering chat tomorrow if my plan violates 1:1:1. If that is the case then we will need to probably do a good amount of work to make a new endpoint and make it 1:1:1.

**Comment by blimpich** [comment]
Seems like we can't use my proof of concept branch since the user is doing 1 action, and therefore we should be doing 1 api command. Which means we need to probably create a new command since the current `AddBillingCardAndRequestPolicyOwnerChange` is not 1:1:1. Will start work on that.

**Comment by blimpich** [comment]
Well actually, the more I think about this the more I think its literally not possible to do this all in 1 api command. We have to go through stripe 3ds verification process, which requires first getting an authentication link from stripe in the backend. And then with that authentication link we render a new 3rd party screen that the user uses to verify their payment method. I don't know how we would do that without making two calls. Will ask engineering-chat

**Comment by blimpich** [comment]
Yeah this is a weird edge case of 1:1:1 since if the card doesn't need SCA authentication then technically we are breaking 1:1:1, but if it does need it then the two api commands actually kind of make sense since the user does 1 action to add the card, and then one action to authenticate the card.

**Comment by blimpich** [comment]
Okay so I've noodled on this long enough that I'm confident this can be done all on the frontend and in a way that is also 1:1:1 compliant. Here are the flows that I want to happen:

<details><summary>Setup for either flow:</summary>

1. create user A with a collect workspace
2. create user B
3. Add user B to the workspace as an admin
4. user B navigates to Workspaces -> Workspace -> Members -> User A -> click Transfer Owner
5. In the network tab there should be a request to `RequestWorkspaceOwnerChange` that responds with `jsonMessage: "Current user does not have a billing card."`
</details>

<details><summary>Flow 1: Transfer ownership with a non-GBP billing card</summary>

1. Fill out payment card form and make sure the currency **is not** `GBP`.
2. Click "Accept terms & add payment card"
3. In the network tab we should see 1 call to `AddPaymentCard` and the UI should change to a right-pane that says "Take over annual subscription". See below image for reference.
![Screenshot 2024-08-06 at 4 41 58 PM](https://github.com/user-attachments/assets/5ac5ef52-4c02-4df3-9a89-3ac69971c277)

4. Click "Transfer subscription" button
5. In the network tab should see 1 new call to `RequestWorkspaceOwnerChange` and should see success screen that says "Woohoo! All set."

_Note: this is very similar to how the current implementation works, except instead of calling `AddPaymentCard` we call `AddBillingCardAndRequestPolicyOwnerChange`._

</details>

<details><summary>Flow 2: Transfer ownership with a GBP billing card</summary>

1. Fill out payment card form and make sure the currency is `GBP`.
2. Click "Accept terms & add payment card"
3. In the network tab we should see 1 call to `AddPaymentCardGBP` and a new iframe modal should pop up with Stripe's 3DS verification screen.
4. Click "complete"
5. In the network tab there should be 1 call to `User_VerifySetupIntent`, the iframe modal should close, and the right-pane UI should change to the UI that says "Take over annual subscription".
6. Click "Transfer subscription" button
7. In the network tab should see 1 new call to `RequestWorkspaceOwnerChange` and should see success screen that says "Woohoo! All set."

</details>

Completing this will probably involve refactoring how our current transfer owner code is setup and also will need to leverage [code that was added](https://github.com/Expensify/App/pull/44321) recently to make [Strong Customer Authentication (SCA)](https://stripe.com/guides/strong-customer-authentication) enabled billing cards possible.

In order to work on 3DS authentication you'll also need to use my ngrok credentials, which are listed below.

<details><summary>Ngrok credentials</summary>

```
NEW_EXPENSIFY_URL=https:/
NEW_EXPENSIFY_URL=https://new.expensify.com/
SECURE_EXPENSIFY_URL=https://secure.expensify.com.dev/
EXPENSIFY_URL=https://www.expensify.com.dev/
EXPENSIFY_PARTNER_NAME=chat-expensify-com
EXPENSIFY_PARTNER_PASSWORD=e21965746fd75f82bb66
PUSHER_APP_KEY=ac6d22b891daae55283a
SECURE_NGROK_URL=https://secure-expensify-blimpich.ngrok.io/
NGROK_URL=https://expensify-blimpich.ngrok.io/
USE_NGROK=true
USE_WEB_PROXY=false
USE_WDYR=false
CAPTURE_METRICS=false
ONYX_METRICS=false

PUSHER_DEV_SUFFIX=-def525f225124b848df43bee41aa4e7c
```

Notes on using my ngrok credentials:
1. all magic codes will be `000000`
2. you won't get most emails/notifications unless I run a script on my local machine
3. my ngrok will only work when I'm online. I generally work 9am-5pm PST hours.
</details>

**Comment by blimpich** [comment]
The problem with this plan is that the current frontend is designed around redundant calls to transferPolicyOwnership in the backend. Which means this is actually _a lot_ of work if we actually want this to be actually 1:1:1.

**Comment by blimpich** [comment]
Ignore the ping @RachCHopkins, I thought I was gonna make this external but realized I need to do some backend work first.

**Comment by blimpich** [comment]
Putting this on hold since we need the backend changes first before a contributor can pick this up.

**Comment by blimpich** [comment]
Putting this as weekly since its on hold for now.

**Comment by blimpich** [comment]
Not overdue, on hold.

**Comment by blimpich** [comment]
Updated the description of the issue with new information. Still waiting on held issue to move forward on this.

**Comment by blimpich** [comment]
Intending to assign to @waterim based off [this conversation](https://expensify.slack.com/archives/C036QM0SLJK/p1725574302897319).

**Comment by waterim** [comment]
Hello, Im Artem from Callstack and would like to help with this issue!

**Comment by trjExpensify** [comment]
Done!

**Comment by blimpich** [comment]
Backend PRs are merged, waiting deploy. @waterim if you use my ngrok to start working on this if you should be able to hit the new `VerifySetupIntentAndRequestPolicyOwnerChange` endpoint. If not then just wait till tomorrow when this should get deployed and you'll be able to hit the endpoint without using my ngrok.

**Comment by blimpich** [comment]
Though I suppose you'll need my ngrok anyways to get SCA to work, forgot about that 😅

**Comment by waterim** [comment]
@blimpich Thank you!
Starting today with that issue

**Comment by JmillsExpensify** [comment]
Looks like we're still working through the PR

**Comment by waterim** [comment]
Not overdue, under development
Im in contact with @blimpich

**Comment by blimpich** [comment]
PR is in review, @ntdiary you will need my ngrok credentials in order to test this PR. Please see [this comment](https://github.com/Expensify/App/issues/42490#issuecomment-2272376992) for the credentials and some common questions about using ngrok. Please put those credentials into your `.env` file in App.

**Comment by blimpich** [comment]
Reassigning C+ as per [this discussion](https://expensify.slack.com/archives/C02NK2DQWUX/p1727251073758199).

@DylanDylann please see this comment for my ngrok credentials and details on the flows that we are testing: https://github.com/Expensify/App/issues/42490#issuecomment-2272376992

**Comment by blimpich** [comment]
@DylanDylann I'm signing off for the day but are you going to be able to review this tomorrow during my work hours (9am-5pm PST)? I'd like to keep the ball rolling on this PR. Thank you!

**Comment by DylanDylann** [comment]
@JmillsExpensify The PR is deployed to production on 2/10. The payment should be ready on 9/10

**Comment by JmillsExpensify** [comment]
Ah sorry about that. 

Payment summary: $250 for @DylanDylann for PR review and testing. Payment via Upwork.

**Comment by JmillsExpensify** [comment]
All paid out.
